lang/java/avro/src/main/java/org/apache/avro/util/internal/JacksonUtils.java

```diff
@@ -136,10 +137,23 @@ public static Object toObject(JsonNode jsonNode, Schema schema) {
         return jsonNode.asDouble();
       }
     } else if (jsonNode.isDouble() || jsonNode.isFloat()) {
-      if (schema == null || schema.getType().equals(Schema.Type.DOUBLE)) {
-        return jsonNode.asDouble();
-      } else if (schema.getType().equals(Schema.Type.FLOAT)) {
-        return (float) jsonNode.asDouble();
+      if (schema != null) {
+        if (schema.getType().equals(Schema.Type.DOUBLE)) {
+          return jsonNode.doubleValue();
+        } else if (schema.getType().equals(Schema.Type.FLOAT)) {
+          return jsonNode.floatValue();
+        }
```

Comment:
Is it possible to have a JSON node with a float/double, but a schema that is neither?

Reply:
if there is an error in the data, i guess it's possible.